src/seq_analysis/orf.rs

```diff
@@ -0,0 +1,163 @@
+//!    let orf = &sequence[start..end];
+//!    //...do something with orf sequence...
+//! }
+//! ```
+//!
+//! Right now the only way to check the reverse strand for orf is to use
+//! the alphabet::dna::RevComp struct and to check for both sequences.
+//! But that's not so performance friendly, as the reverse complementation and the orf research
+//! could go on at the same time.
+
+
+use utils::{TextIterator, IntoTextIterator};
+use std::collections::VecDeque;
+use std::iter;
+
+
+/// An implementation of a naive algorithm finder
+pub struct Finder {
+    start_codons: Vec<VecDeque<u8>>,
+    stop_codons: Vec<VecDeque<u8>>,
+    min_len: usize,
+}
+
+impl Finder {
+    /// Create a new instance of a finder for the given start and stop codons and a particular length
+    pub fn new<'a>(start_codons: Vec<&'a [u8; 3]>,
+                   stop_codons: Vec<&'a [u8; 3]>,
+                   min_len: usize)
+                   -> Self {
+        Finder {
+            start_codons: start_codons.into_iter()                          // Convert start_ and
+                                      .map(|x| {                            // stop_codons from 
+                                          x.into_iter()                     // Vec<&[u8;3]> to
+                                           .map(|&x| x as u8)               // Vec<VecDeque<u8>>
+                                           .collect::<VecDeque<u8>>()       // so they can be
+                                      })                                    // easily compared
+                                      .collect(), // with codon built
+            stop_codons: stop_codons.into_iter()                            // from IntoTextIterator
+                                    .map(|x| {                              // object.
+                                        x.into_iter()
+                                         .map(|&x| x as u8)
+                                         .collect::<VecDeque<u8>>()
+                                    })
+                                    .collect(),
+            min_len: min_len,
+        }
+    }
+
+    /// Find all orfs in the given sequence
+    pub fn find_all<'a, I: IntoTextIterator<'a>>(&'a self, seq: I) -> Matches<I::IntoIter> {
+        Matches {
+            finder: self,
+            state: State::new(),
+            seq: seq.into_iter().enumerate(),
+        }
+    }
+}
+
+/// An orf representation with start and end position of said orf,
+/// as well as offset of the reading frame (1,2,3) and strand location 
+// (current: +, reverse complementary: -).
+pub struct Orf {
+    pub start: usize,
+    pub end: usize,
+    pub offset: i8,
+}
+
+/// The current algorithm state.
+struct State {
+    start_pos: [usize; 3],
+    in_orf: [bool; 3],
+    codon: VecDeque<u8>,
+}
+
+
+impl State {
+    /// Create new state.
+    pub fn new() -> Self {
+        State {
+            start_pos: [0, 0, 0],
+            in_orf: [false, false, false],
```

Comment:
Done !
Le 21 mai 2016 7:36 PM, "Johannes Köster" <notifications@github.com> a
écrit :

> In src/seq_analysis/orf.rs
> <https://github.com/rust-bio/rust-bio/pull/48#discussion_r64137711>:
>
> > +}
> > +
> > +/// The current algorithm state.
> > +struct State {
> > +    start_pos: [usize; 3],
> > +    in_orf: [bool; 3],
> > +    codon: VecDeque<u8>,
> > +}
> > +
> > +
> > +impl State {
> > +    /// Create new state.
> > +    pub fn new() -> Self {
> > +        State {
> > +            start_pos: [0, 0, 0],
> > +            in_orf: [false, false, false],
>
> One final issue, sorry! I think it would be more rustic to write start_pos:
> [None, None, None]. Then, you can insert start positions with Some(0),
> and replace the if else below with a match construct.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-bio/rust-bio/pull/48/files/2e4800f3ffe5ed6f0a6b89d951ef691ea6c842c6#r64137711>
>